File: 0x08/BOJ10799.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    string s;
    cin >> s;

    bool wasPoped = false;  // 이전에 레이저였는지
    stack<char> st;
    int result = 0;

    for(auto a : s) {
        if(a=='(') {
            st.push(a);
            wasPoped = false;
        } else {  //  ')' 인 경우
            if(!st.empty()) {
                if(!wasPoped && st.top()=='(') {  // 레이저인 경우
                   st.pop();
                   result += st.size();
                   wasPoped = true;
                } else if (wasPoped) {  // 짝 찾은 경우
                   st.pop();
                   result++;
                   wasPoped = true; 
                }
            }
        }
    }

    cout << result;
}

File: barking_workbook/0x08/stack/BOJ10799.java
package stack;

import java.util.*;

public class BOJ10799 {
    public static void main (String[] args) throws Exception {

        Scanner input = new Scanner(System.in);

        String s = input.next();
        Stack<Character> st = new Stack<>();

        boolean wasPoped = false;
        int result = 0;

        for(int i = 0; i < s.length(); i++) {
            if(s.charAt(i) == '(') {
                st.push(s.charAt(i));
                wasPoped = false;
            } else {    // ')' 이면
                if(!st.empty() && st.peek() == '(') {
                    st.pop();
                    
                    // 1. 레이저인 경우
                    if(!wasPoped) result += st.size();
                    // 2. 짝 찾은 경우
                    else result++;

                    wasPoped = true;
                }
            }
        }

        System.out.println(result);

        input.close();

    }
}